### PR TITLE
Fix #5225: Add "Debouncing"

### LINKED ISSUE #5225: Add debouncing to iOS
### Description: 
In order to stop trackers we skip certain middle man url and go directly to the redirected urls. This is called debouncing. More [here](https://brave.com/privacy-updates/11-debouncing/).

The trackers to debounce could be found here:
https://github.com/brave/adblock-lists/blob/master/brave-lists/debounce.json

### Testing
The following site has been provided for testing debouncing:
https://dev-pages.brave.software/navigation-tracking/index.html
However, this does not support the testing of `base64` decoded redirect URLs

### Related
https://github.com/brave/devops/issues/7220

### PR DESCRIPTION
**Pull request by cuba**
## Summary of Changes

This pull request fixes #5225 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()` (N/A)

## Test Plan:

A couple of tests are available here:
https://dev-pages.brave.software/navigation-tracking/index.html
and here:
https://fingerprint-protection-test.herokuapp.com/debounce

In addition, tests have been added to cover a wide variety of scenarios and based on the tests in `brave-core`

## Screenshots:

N/A

## Linked
https://github.com/brave/security/issues/825

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Shared/Extensions/URLExtensions.swift
Comment (kylehickinson) [inline Shared/Extensions/URLExtensions.swift]: FYI this API is being deprecated in this PR: https://github.com/brave/brave-ios/pull/5239/files

Comment (cuba) [reply Shared/Extensions/URLExtensions.swift]: > FYI this API is being deprecated in this PR: https://github.com/brave/brave-ios/pull/5239/files

I will need access to be able to extract the eTLD+1 from a host string, not a URL. Do you know if I can do that with the new API?

Comment (kylehickinson) [reply Shared/Extensions/URLExtensions.swift]: As far as I know, no, the Chromium API is based on URLs not strings. Is it not feasible to make a URL from the host?

cc @Brandon-T

Comment (cuba) [reply Shared/Extensions/URLExtensions.swift]: Ok, I reverted my URLExtension. For now I create a temporary url like `//youtube.com` to extract the ETLD and I will adjust it according to Brandon's feedback.

**Comment by cuba** [comment]
@kylehickinson @iccub FYI I just modied the PR. I forgot to add the startLoading invocation on the scene delegate and to check for a first time load (so I'm not pointlessly re-loading cached data)

File: Client/WebFilters/DebouncingResourceDownloader.swift
Comment (kylehickinson) [inline Client/WebFilters/DebouncingResourceDownloader.swift]: I think these exist already unless I'm missing some difference? https://github.com/brave/brave-ios/blob/f268386428cbe308ad6a479f82d115ce29d755ee/Client/Extensions/FileManagerExtension.swift#L76

**Review by kylehickinson — CHANGES_REQUESTED** [review]
re: duplicated file manager extension methods

**Comment by cuba** [comment]
> re: duplicated file manager extension methods

@kylehickinson I removed my get or create directory implementation and using the existing one.
cc @Brandon-T 

Other changes I made were this:
1. I implemented the same tests that can be found in brave-core. This required some slight modifications to the algorithm I had: Allow handling of redirect in redirect in redirect etc scenarios.
2. I strip out everything from the modified request (I essentially make a new one) except for `Referer`. I'm still not totally sure about this.
3. only handle `http` and `https` scenarios (as is in brave-core)

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (Brandon-T) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: What happens to the `Cookies` and `X-Session` and other headers that are usually passed along with a request?
For example, when logging into certain sites (like facebook), it might have an `Is this you` login button that you can click and it pulls the cookies, a token, and session from storage and sends it in the request.

Will this affect that?

File: Client/WebFilters/DebouncingResourceDownloader.swift
Comment (Brandon-T) [inline Client/WebFilters/DebouncingResourceDownloader.swift:136]: Is there any such pattern where `redirectURLOnce` will call `redirectURLRecursively` infinitely or enough times to blow the stack?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Comment by Brandon-T** [comment]
Any PRs modifying requests, cookies, etc. must have a Security Review. 

You can open one here: https://github.com/brave/security/issues/new?assignees=&labels=security%2C+security-reviews%2C+privacy-reviews%2C+priority%2FP4&template=security-review.md&title=%5BiOS%5D%20Security%2FPrivacy+Review+for+Request+Debouncing+%5BJacob%20Sikorski%5D

Link the PR and possible Spec (or N/A if none available).
Then edit this PR to link back to the opened Sec Review :)

**Comment by cuba** [comment]
> Any PRs modifying requests, cookies, etc. must have a Security Review.
> 
> You can open one here: https://github.com/brave/security/issues/new?assignees=&labels=security%2C+security-reviews%2C+privacy-reviews%2C+priority%2FP4&template=security-review.md&title=%5BiOS%5D%20Security%2FPrivacy+Review+for+Request+Debouncing+%5BJacob%20Sikorski%5D
> 
> Link the PR and possible Spec (or N/A if none available). Then edit this PR to link back to the opened Sec Review :)

Done!

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (fmarier) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:201]: I'd suggest making sure that the redirect URL's scheme is HTTP or HTTPS before debouncing it. The list comes from us and should be okay, but it's not a bad thing to be extra-cautious here.

**Review by Brandon-T — APPROVED** [review]

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:201]: You can do `url.isWebPage(includeDataURIs: false)` for this btw

**Review by kylehickinson — APPROVED** [review]

**Comment by kylehickinson** [comment]
Waiting on sec review

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (pes10k) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: nit: is this a swift convention, to do variable assignment a) on the same line, and b) w/in a conditional? 

If its a convention, then understood, but to a non-swift person, this line is difficult to parse for me and seems like a possible area for confusion

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (pes10k) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: Is this still correct?  its only Referer now we target, and the system handles everything else, right?

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (pes10k) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: if the set is always of size one, couldn't this (and the below loop) be rewritten to just check for `Refer` and simplify all this?

File: Client/WebFilters/DebouncingResourceDownloader.swift
Comment (pes10k) [inline Client/WebFilters/DebouncingResourceDownloader.swift:156]: nit : why `Once`?  I dont see any logic in the function that is checking to make sure something only happens once.  Am I missing something?  Otherwise this seems like an opportunity for confusion

**Comment by cuba** [comment]
> Waiting on sec review

@kylehickinson Sec ticket is closed now. Could we remove the block?

**Comment by kylehickinson** [comment]
Can be merged once CI finishes, please choose "Squash and merge" when merging :)

**Pull request merged**
